File: src/javax/microedition/lcdui/GameDimensions.java
package javax.microedition.lcdui;

/**
 * Logical framebuffer size for the desktop MIDP stack and AquaDig.
 *
 * <p>Fixed mode (default): logical size {@link #FIXED_WIDTH}×{@link #FIXED_HEIGHT}, window size matches
 * the 2× scaled buffer (see {@link Canvas}).
 *
 * <p>Dynamic mode: the window is resizable; logical size tracks the canvas panel in <em>physical</em>
 * pixels divided by {@link #BUFFER_SCALE} (same 2× upscale as fixed mode: each logical pixel is drawn
 * as a 2×2 block).
 *
 * <p>Enable resizable / dynamic logical size with JVM flag {@code -Daquadig.dynamicResolution=true}
 * or program argument {@code --dynamic-resolution} before starting the MIDlet (see {@link aquadig.AquaDig}).
 */
public final class GameDimensions {
  /** Logical → framebuffer pixels (matches {@link Canvas} scaling). */
  public static final int BUFFER_SCALE = 2;

  /** Default logical size: classic MIDP portrait (2× buffer in {@link Canvas}). */
  public static final int FIXED_WIDTH = 400;
  public static final int FIXED_HEIGHT = 300;

  /** @deprecated use {@link #FIXED_WIDTH} */
  public static final int WIDTH = FIXED_WIDTH;

  /** @deprecated use {@link #FIXED_HEIGHT} */
  public static final int HEIGHT = FIXED_HEIGHT;

  private static volatile boolean dynamic =
      Boolean.parseBoolean(System.getProperty("aquadig.dynamicResolution", "false"));

  private static volatile int runtimeWidth = FIXED_WIDTH;
  private static volatile int runtimeHeight = FIXED_HEIGHT;

  private GameDimensions() {}

  public static boolean isDynamic() {
    return dynamic;
  }

  public static void setDynamic(boolean v) {
    dynamic = v;
  }

  /** Logical width in game/MIDP coordinates (pixels). */
  public static int getWidth() {
    return dynamic ? runtimeWidth : FIXED_WIDTH;
  }

  /** Logical height in game/MIDP coordinates (pixels). */
  public static int getHeight() {
    return dynamic ? runtimeHeight : FIXED_HEIGHT;
  }

  /**
   * Updates runtime logical size from the panel's physical size (dynamic mode). Logical width/height
   * are {@code physical / BUFFER_SCALE} so rendering stays 2× scaled like fixed mode.
   *
   * @return true if logical width or height changed
   */
  public static boolean setRuntimeLogicalSizeIfChanged(int physicalWidth, int physicalHeight) {
    if (!dynamic) {
      return false;
    }
    int lw = Math.max(1, physicalWidth / BUFFER_SCALE);
    int lh = Math.max(1, physicalHeight / BUFFER_SCALE);
    if (lw == runtimeWidth && lh == runtimeHeight) {
      return false;
    }
    runtimeWidth = lw;
    runtimeHeight = lh;
    return true;
  }

  /** Resets runtime dimensions to fixed defaults (e.g. when toggling dynamic off). */
  public static void resetRuntimeToFixed() {
    runtimeWidth = FIXED_WIDTH;
    runtimeHeight = FIXED_HEIGHT;
  }
}

File: docs/swingjs/j2s/javax/microedition/lcdui/GameDimensions.js
(function(){var P$=Clazz.newPackage("javax.microedition.lcdui"),I$=[];
/*c*/var C$=Clazz.newClass(P$, "GameDimensions");

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
},1);

C$.$fields$=[[]
,['Z',['dynamic'],'I',['runtimeWidth','runtimeHeight']]]

Clazz.newMeth(C$, 'c$',  function () {
;C$.$init$.apply(this);
}, 1);

Clazz.newMeth(C$, 'isDynamic$',  function () {
return C$.dynamic;
}, 1);

Clazz.newMeth(C$, 'setDynamic$Z',  function (v) {
C$.dynamic=v;
}, 1);

Clazz.newMeth(C$, 'getWidth$',  function () {
return C$.dynamic ? C$.runtimeWidth : 400;
}, 1);

Clazz.newMeth(C$, 'getHeight$',  function () {
return C$.dynamic ? C$.runtimeHeight : 300;
}, 1);

Clazz.newMeth(C$, 'setRuntimeLogicalSizeIfChanged$I$I',  function (physicalWidth, physicalHeight) {
if (!C$.dynamic) {
return false;
}var lw=Math.max(1, (physicalWidth/2|0));
var lh=Math.max(1, (physicalHeight/2|0));
if (lw == C$.runtimeWidth && lh == C$.runtimeHeight ) {
return false;
}C$.runtimeWidth=lw;
C$.runtimeHeight=lh;
return true;
}, 1);

Clazz.newMeth(C$, 'resetRuntimeToFixed$',  function () {
C$.runtimeWidth=400;
C$.runtimeHeight=300;
}, 1);

C$.$static$=function(){C$.$static$=0;
C$.dynamic=Boolean.parseBoolean$S(System.getProperty$S$S("aquadig.dynamicResolution", "false"));
C$.runtimeWidth=400;
C$.runtimeHeight=300;
};
})();
;Clazz.setTVer('3.3.1-v5');//Created 2026-06-25 15:29:03 Java2ScriptVisitor version 3.3.1-v5 net.sf.j2s.core.jar version 3.3.1-v5
